### ISSUE TITLE
useStorage() returning undefined before returning value

### ISSUE THREAD

**Issue by zorridge** [PlasmoHQ/storage#50]
**background:**
- using the `useStorage()` hook from a React component via content scripts UI

![image](https://github.com/PlasmoHQ/storage/assets/86993236/689631e9-d36f-474a-bc59-3748dbb83120)

**current behaviour:**
- `undefined` is returned before the actual value, resulting in double renders

![image](https://github.com/PlasmoHQ/storage/assets/86993236/7717d695-209d-4205-aab1-e1ff3dacc203)

**expected behaviour:**
- only actual value is returned

versions:
```
"@plasmohq/storage": "^1.9.0",
"plasmo": "^0.84.1",
react": "18.2.0",
```

**Comment by Acorn221** [comment]
Hi, I'm pretty sure this is expected behavior and that it goes from undefined -> {} when it retrieves the entry from the storage.

**Comment by zorridge** [comment]
do you happen to know what's the rationale behind not awaiting for the value to be retrieved from the storage before returning? I assume it is returning undefined because the `get` method is async

**Comment by apteryxxyz** [comment]
This is expected behaviour, value is `undefined` between when its state is initialised, and the actual value is retrieved from the storage. Unless you pass a default value as the second parameter, in which case instead of `undefined` you will see the default value.

**Comment by louisgv** [comment]
closing as resolved by #53 , #52 , #49

**Comment by SamHoque** [comment]
> Been having lots of problems with this too, my onInit function keeps reseting the default value of my state. I though the onInit function would run only after it was sure it got the value from storage
> 
> ![image](https://private-user-images.githubusercontent.com/74543594/300507886-d79551ba-ba77-4710-bd87-711d196ff473.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQyOTc0MDUsIm5iZiI6MTcxNDI5NzEwNSwicGF0aCI6Ii83NDU0MzU5NC8zMDA1MDc4ODYtZDc5NTUxYmEtYmE3Ny00NzEwLWJkODctNzExZDE5NmZmNDczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI4VDA5MzgyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVkMDFkOGQ0NjA2NjY4MDk5NWM0NDU3NDRjYzRjODM3ZDAxYmQxMTYxN2I4OTQ0YWZkMjdkOGRjZjU5NjhjOTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BWwu6o8g3oZd11aWxgAYUhiYr2eJwa3s5LAZQ3-yJt4)
> 
> would be good to have a way to tell if the underlying value is undefined or if it's just still fetching because its async

Did you figure this out? my useStorage hook shows its default value and then later changes to the stored value, how do we know its still fetching?
